### PR TITLE
fix: preserve the type of failures that are instances of Error

### LINKED ISSUE #12: Bug: type of the error failure is not preserved
When the failure/left type is an instance of an Error class, the original type of the error is not preserved. The plugin wraps the error cause in a new Error class.

## Steps to reproduce

```ts
import SchemaBuilder from '@pothos/core';
import { Effect } from 'effect';
import { execute, parse } from 'graphql';
import EffectPlugin from 'pothos-plugin-effect';

interface SchemaTypes {
  Context: {
    diceResult: number;
  };
}

const builder = new SchemaBuilder<SchemaTypes>({
  plugins: [EffectPlugin],
  relayOptions: {},
});

class CustomError extends Error {
  readonly _tag = 'CustomError';
}

builder.queryField('roll', t =>
  t.effect({
    // @ts-expect-error
    resolve() {
      return Effect.fail(new CustomError("Boom!"));
    },
    type: ['Int'],
  })
);

const schema = builder.toSchema();
const document = parse(`{ roll }`);

execute({ document, schema }).then(console.log, console.error);
```

### Expected behavior

The `pothos-plugin-effect` resolver throws an error of an instance of `CustomError`.

### Actual behavior

The `pothos-plugin-effect` resolver wraps the `CustomError` in an instance of `Error`, or `failErrorConstructor`, or `defaultFailErrorConstructor` when so configured.

I think the current behavior is faulty mostly because the [example code here](https://github.com/iamchanii/pothos-plugin-effect/blob/main/examples/02-with-errors-plugin/src/index.ts) suggests that the expected behavior described above should be the intended behavior, as well.

## Motivation

I'm using the [pothos-errors plugin](https://pothos-graphql.dev/docs/plugins/errors), and I need the custom error class to be preserved for Pothos to be able to return the error as part of the expected error set in the `data` response.

### PR DESCRIPTION
**Pull request by vecerek**
Fixes #12

### CODE REVIEW

**Comment by vecerek** [comment]
@iamchanii what does the following code mean:

```ts
if (Cause.isAnnotatedType(result.cause) && Cause.isFailType(result.cause.cause)) {
  throw result.cause.cause.error;
}
```

and how would I be able to return an annotated type as an alternative to this fix?

**Comment by iamchanii** [comment]
> @iamchanii what does the following code mean:
> 
> ```ts
> if (Cause.isAnnotatedType(result.cause) && Cause.isFailType(result.cause.cause)) {
>   throw result.cause.cause.error;
> }
> ```
> 
> and how would I be able to return an annotated type as an alternative to this fix?

@vecerek At least when I implemented, `Effect.runPromiseExit` provided annotated exit.

```
EffectPrimitiveFailure {
  _tag: 'Failure',
  i0: {
    _tag: 'Annotated',
    cause: { _tag: 'Fail', error: [NotFound] },
    annotation: StackAnnotation {
      stack: [Object],
      seq: 1,
      [Symbol(@effect/io/Cause/StackAnnotation)]: Symbol(@effect/io/Cause/StackAnnotation)
    }
  },
  i1: undefined,
  i2: undefined,
  trace: undefined,
  [Symbol(@effect/io/Effect)]: { _R: [Function: _R], _E: [Function: _E], _A: [Function: _A] }
}
```

But now we got just Failed exit.

```
EffectPrimitiveFailure {
  _tag: 'Failure',
  i0: {
    _tag: 'Fail',
    error: NotFound: User iamchanii2 not found
        at <anonymous> (/Users/iamchanii/dev/pothos-plugin-effect/examples/02-with-errors-plugin/src/GitHub.ts:51:63)
        at e (/Users/iamchanii/dev/pothos-plugin-effect/node_modules/.pnpm/@effect+io@0.37.1/node_modules/@effect/io/src/internal/effect.ts:250:14)
        at cause (/Users/iamchanii/dev/pothos-plugin-effect/node_modules/.pnpm/@effect+io@0.37.1/node_modules/@effect/io/src/internal/core.ts:683:16)
        at FiberRuntime.Failure (/Users/iamchanii/dev/pothos-plugin-effect/node_modules/.pnpm/@effect+io@0.37.1/node_modules/@effect/io/src/internal/fiberRuntime.ts:1082:25)
        at <anonymous> (/Users/iamchanii/dev/pothos-plugin-effect/node_modules/.pnpm/@effect+io@0.37.1/node_modules/@effect/io/src/internal/fiberRuntime.ts:1278:51)
        at onRun (/Users/iamchanii/dev/pothos-plugin-effect/node_modules/.pnpm/@effect+io@0.37.1/node_modules/@effect/io/src/internal/supervisor.ts:227:12)
        at runLoop (/Users/iamchanii/dev/pothos-plugin-effect/node_modules/.pnpm/@effect+io@0.37.1/node_modules/@effect/io/src/internal/fiberRuntime.ts:1276:32)
        at evaluateEffect (/Users/iamchanii/dev/pothos-plugin-effect/node_modules/.pnpm/@effect+io@0.37.1/node_modules/@effect/io/src/internal/fiberRuntime.ts:846:29)
        at evaluateMessageWhileSuspended (/Users/iamchanii/dev/pothos-plugin-effect/node_modules/.pnpm/@effect+io@0.37.1/node_modules/@effect/io/src/internal/fiberRuntime.ts:813:14)
        at drainQueueOnCurrentThread (/Users/iamchanii/dev/pothos-plugin-effect/node_modules/.pnpm/@effect+io@0.37.1/node_modules/@effect/io/src/internal/fiberRuntime.ts:569:18) {
      _tag: 'NotFound'
    }
  },
  i1: undefined,
  i2: undefined,
  trace: undefined,
  [Symbol(@effect/io/Effect)]: { _R: [Function: _R], _E: [Function: _E], _A: [Function: _A] }
}
```

It means `Effect.runPromiseExit`'s internal implementation in `@effect/io` was changed from `v0.30.0` to now.

File: src/field-builder.ts [hunk 1]
```diff
@@ -63,7 +63,13 @@ fieldBuilderProto.effect = function effect({ effect = {}, resolve, ...options })
...
+      // fixme: result should not be Exit.Exit<never, any>
+      // because never cannot be checked if it's an instanceof Error
       if (Exit.isFailure(result)) {
+        if (Cause.isFailType(result.cause) && (result.cause.error as any instanceof Error)) {
+          throw result.cause.error;
+        }
+
         if (Cause.isAnnotatedType(result.cause) && Cause.isFailType(result.cause.cause)) {
           throw result.cause.cause.error;
         }
```
Comment (iamchanii) [inline src/field-builder.ts:75]: From https://github.com/iamchanii/pothos-plugin-effect/pull/13#issuecomment-1678665201, I think `Cause.isAnnotatedType(result.cause)` is not valid logic for now. 

```suggestion
        if (Cause.isFailType(result.cause) && (result.cause.error as any instanceof Error)) {
          throw result.cause.error;
        }
```

File: src/field-builder.ts [hunk 2]
```diff
@@ -63,7 +63,13 @@ fieldBuilderProto.effect = function effect({ effect = {}, resolve, ...options })
...
       }
 
+      // fixme: result should not be Exit.Exit<never, any>
+      // because never cannot be checked if it's an instanceof Error
       if (Exit.isFailure(result)) {
+        if (Cause.isFailType(result.cause) && (result.cause.error as any instanceof Error)) {
+          throw result.cause.error;
+        }
```
Comment (iamchanii) [inline src/field-builder.ts:71]: `result.cause.error` could can be not an instance of `Error`. so it should be wraped by `failErrorConstructor`, `effectOptions.defaultFailErrorConstructor` or `Error`.

**Review by iamchanii — CHANGES_REQUESTED** [review]

File: src/field-builder.ts (hunk 2 above)
Comment (vecerek) [reply src/field-builder.ts]: I might be misunderstanding you but that's exactly the reason for which I'm also checking if `result.cause.error instanceof Error` in the above code. If it isn't, I want it to fall through and be handled by the throw logic below:

```ts
throw new (effect.failErrorConstructor ?? effectOptions?.defaultFailErrorConstructor ?? Error)(
  Cause.pretty(result.cause),
);
```

**Review by iamchanii — APPROVED** [review]

File: src/field-builder.ts (hunk 2 above)
Comment (iamchanii) [reply src/field-builder.ts]: ahh yep. you're right.

**Pull request merged**